### ISSUE TITLE
Frameless + Drag Handlers Implementation

### ISSUE THREAD

**Issue by mattkol** [chromelyapps/Chromely#96]
@mattkol The drag handler was not added in CefGlue as the System.Drawing library has missing classes. The implementation isn't considered complete for CefGlue. If an alternative can be used to add the necessary classes then it can be done.

I also looked into instead of hard coding the margins (at least for WinApi) to use GetSystemMetrics on the SM_CYFRAME and SM_CXFRAME [see here](https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-getsystemmetrics) for the C++. This API isn't DPI aware so instead it might be best to use GetSystemMetricsForDpi [see here](https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-getsystemmetricsfordpi)

_Originally posted by @NovusTheory in https://github.com/chromelyapps/Chromely/issues/93#issuecomment-491896264_

**Comment by mattkol** [comment]
@NovusTheory @FrankPfattheicher I added the drag handlers - https://github.com/chromelyapps/Chromely/commit/d4486da394a6297f711fe2a230c63e334b6cceb4. 
CefGlue drag handler is just a place holder for now - [CefGlueDragHandler.cs](https://github.com/chromelyapps/Chromely/blob/master/src/Chromely.CefGlue/Browser/Handlers/CefGlueDragHandler.cs) until we find solution with System.Drawing Netstandard implementation.

**Comment by NovusTheory** [comment]
I worked on this a bit and noticed that you actually broke frameless in your adjustments with this https://github.com/chromelyapps/Chromely/commit/44f8abee7053e3b255e3dda8b040bb23fd07ee47#diff-099ca516144c41580f77cc3e65775029L186

This is necessary to use the custom drag regions and added them back in my commit (view below).

I won't pull request anything yet but in my commit (https://github.com/NovusTheory/Chromely/commit/6dd0661f937c264b6aed9386fe7f9aa0e96e7888) I fixed the frameless window mode systems and added it to CefGlue now. I noticed that System.Drawing.Common works and added that NuGet package which adds the Region class. Everything in this commit works and have tested with your demo applications respectfully. In an old issue #93 it said it had an issue with mouse clicks but I have not experienced this at all with and without frameless mode enabled.

If you'd like to review it on the spot before I send a PR then you may.

**Comment by mattkol** [comment]
@NovusTheory that is not totally unexpected that something broke - reason I asked for a review - https://github.com/chromelyapps/Chromely/issues/93#issuecomment-491525455

I may be able to review by weekend. @FrankPfattheicher  is also familiar with this so he may have some bandwidth for a review.

Thanks.

**Comment by NovusTheory** [comment]
I'm sorry I never actually got around to fully reviewing that commit and completely let it pass.

**Comment by mattkol** [comment]
@NovusTheory np. Thanks for keeping a track on this ..
We should get it right this time.

**Comment by FrankPfattheicher** [comment]
@NovusTheory 
Code looks good - but there is no content any more !?

Running Tests/Chromely.Integration.TestApp should look like this:
(adding .WithFramelessHost() and commenting out Environment.Exit(0);)
![image](https://user-images.githubusercontent.com/921532/60240286-bfbcd380-98b0-11e9-89db-9b768670d5ff.png)

But i now see an empty window :-(

DevTools show that document is sized to 1016x17 pixel !
Please look at this...

**Comment by mattkol** [comment]
@NovusTheory @FrankPfattheicher is it advisable to include the PR - https://github.com/chromelyapps/Chromely/pull/104/commits/c86ca10511015916958685fafaf254b236a2effb in this test?

If you think it is, please include.

Thanks.

**Comment by FrankPfattheicher** [comment]
@NovusTheory @mattkol 
I think it's better to first fix the "content missing" issue (see above).

**Comment by NovusTheory** [comment]
@FrankPfattheicher I'll run some tests to see why it does this. On a side note (maybe this needs another issue opened). I was doing some tests with GTK on Windows and notice the browser window isn't resized with the parent

**Comment by mattkol** [comment]
@NovusTheory I tested your PR #107 extensively and I noticed 2 main things:

- CefGlue -  draggable does not work but all functionalities work as expected.
- CefSharp - draggable works, but most functionalities are not working or too slow. Sometimes I had to click a button twice.

When I localized it seems to be an issue with a combination of subclassing and event loop. 

My suggestion is to hit a reset on this. May be start incremental changes and we all @NovusTheory @FrankPfattheicher @mattkol  test before we move to the next stage.

Meanwhile for that reset - I had the base frameless + draggable working by combining @NovusTheory [DwmExtendFrame   implementation](https://github.com/chromelyapps/Chromely/blob/ebdeb3edc569ed34aa5f8a199581a8de0d5d2817/src/Chromely.CefSharp.Winapi/BrowserWindow/NativeWindow.cs#L292) and [NC HIT implementation  I found](https://github.com/chromelyapps/Chromely/blob/ebdeb3edc569ed34aa5f8a199581a8de0d5d2817/src/Chromely.CefSharp.Winapi/BrowserWindow/NativeWindow.cs#L329). 

It is frameless and draggable but not re-sizable. Changeset - https://github.com/chromelyapps/Chromely/commit/ebdeb3edc569ed34aa5f8a199581a8de0d5d2817

If this is what is intended from the beginning then we may be good. If not we may take it one step at a time, to find out where the disconnect is. May be we do separately - event looping, subclassing, message handling. But ensuring one step works before the next stage.

@NovusTheory @FrankPfattheicher please also help review this change before we move to the next.
Note that this base implementation does not use subclassing, draggable handler, NC related message handling. We can start re adding these functionalities as needed going forward.

Thanks.

![image](https://user-images.githubusercontent.com/18384207/60758668-82a6be80-9fdf-11e9-9c48-0cd4a4c6d927.png)

![image](https://user-images.githubusercontent.com/18384207/60758675-918d7100-9fdf-11e9-8da9-a16d316557aa.png)

![image](https://user-images.githubusercontent.com/18384207/60758679-9eaa6000-9fdf-11e9-8190-9a6c845a7562.png)

**Comment by mattkol** [comment]
> @NovusTheory @mattkol
> I think it's better to first fix the "content missing" issue (see above).

This is what I get with Integration Test: 

![image](https://user-images.githubusercontent.com/18384207/60758894-4fb1fa00-9fe2-11e9-866a-70f4e97f1e3b.png)

**Comment by NovusTheory** [comment]
I'm not sure what specifically is going on with the two because I've done my own testing and they worked just fine.

> It is frameless and draggable but not re-sizable. Changeset - https://github.com/chromelyapps/Chromely/commit/ebdeb3edc569ed34aa5f8a199581a8de0d5d2817

As for this, the intention is to support this for replicating Electron's frameless window which removes the titlebar but keeps resize functionality as well as the ability to implement custom drag areas with `-webkit-app-region: drag`. Look at some apps in Electron, for example Discord, which does exactly what is being replicated. It uses a custom titlebar with a defined `-webkit-app-region: drag` on it to permit dragging the window around as well as still having the resize borders on the window. It acts as if it was a native window.

If you'd like to hit a restart on this we can figure something out that makes this more clean and take it step by step.

**Comment by mattkol** [comment]
@NovusTheory I will be honest with you, I am confused too. Especially why both CefSharp and CefGlue would not behave the same way. I think you and @FrankPfattheicher have done everything right. 

But please trying test the base implementation that I have, then we can start from there. I think we have all functionalities we just need to combine them correctly.

**Comment by NovusTheory** [comment]
@mattkol I've looked at your base implementation and while it does add the dragging system it's using the native titlebar which isn't the idea as that's not a true frameless window. What is required to keep moving forward is implementing NCHITTEST correctly which allows you to override the "tests" that Windows is performing to determine the action allowed, such as dragging, resizing, and more, for the specific point on screen the mouse cursor is at.

The problem with this override when I was making it was chromium because it's taking the NCHITTEST and providing its own response (by default is no action). This could be where it got finnicky in the code. I had to override the WndProc of chromium's windows and check the hit test (for our non client area which were the resize borders) and bubble back up to the main process (Which is our Chromely app) to allow it to provide a proper NCHITTEST response. Note that we can't provide a response from the overriden WndProc as it would apply to the window which returns that response (not our Chromely app).

The next problem is supporting `-webkit-app-region`. We have to override chromium's LBUTTONDOWN and check the mouse cursor point to see if it's within our defined app-region. If it is then we release the capture and send a message to the main window (our Chromely app) to start dragging.

The reason we have to bubble the events up and not overriding and returning our own response is because Windows would then perform that action to the window that returned that response. In this case it would be the chromium browser and we'll instead be interacting with that which would move it around inside our chromely app.

**Comment by FrankPfattheicher** [comment]
@mattkol @NovusTheory 
What is the current state of th problem?
Can i do something?

It still looks like this:
![image](https://user-images.githubusercontent.com/921532/61366560-169f4280-a88a-11e9-85ff-d8191b0abea2.png)

**Comment by mattkol** [comment]
@FrankPfattheicher yes that is correct but should be draggable. But the drag is using DwmExtend Frame implementation. 

Yes please do something. Thanks.

**Comment by NovusTheory** [comment]
*Disclaimer: I am not looking at CefSharp specifically for this comment.*

I was doing some research and it seems the best way for us to create a frameless window with all the fancy features is doing what views do in CEF. A related issue from [on CEF's bitbucket here](https://bitbucket.org/chromiumembedded/cef/issues/1749/).

Quoted from the issue linked
>>>
    Ownership of the complete window hierarchy including desktop/system interaction, parent/child relationships of windows and widgets, focus/keyboard handling, drag/drop handling, modality, rendering, etc.
    Layout managers (box, fill, grid, etc).
    Widgets (web view, button, text field, bubble, tooltip, table, tree, combo box, etc).
>>>

I'm not sure how useful this is but the most important seems to be this part here
>Ownership of the complete window hierarchy including desktop/system interaction, parent/child relationships of windows and widgets, focus/keyboard handling, drag/drop handling, modality, rendering, etc.

Another relevant file is [CEF's own implementation here](https://bitbucket.org/chromiumembedded/cef/src/master/libcef/browser/views/window_view.cc) which seems to show that it is in control of the window and rather instead passes events down to the browser window [like shown with ShouldDescendIntoChildForEventHandling](https://bitbucket.org/chromiumembedded/cef/src/15fc7e58a880c94c891faf6aa193303f0302d006/libcef/browser/views/window_view.cc#lines-421).

To sum up it seems like we need to either create our own system that works like this or rather implement the views system and rely on CEF to generate windows as it seems to be doing with the views system and is cross platform compatible supposedly.

**Comment by mattkol** [comment]
@NovusTheory personally, my concern with the approach is a boundary issue. Chromely is based on CefSharp and CefGlue implementation of CEF. If these functionalities are not exposed by either implementations, are we not crossing into CEF boundary? There is no problem with that,  but we may require more knowledge to do that successfully and we need to think of maintenance too.

But if that is not the case, then we can explore more.

**Comment by NovusTheory** [comment]
Both these implementations of CEF are just wrappers over the the native CEF DLL. Views are not exposed by CefGlue as far as I can tell and I tried to look for any updates and it didn't seem to include them. I did not look at CefSharp as per my disclaimer in the top of my comment, I was looking at the cross-platform solution more.

Since as far as I can tell CefGlue does not expose the views implementation of CEF so we would end up crossing the boundary of having to maintain that part ourselves. I've forked and looked at CefGlue to see how implementing it would go and it doesn't look all that difficult but I noticed that CefGlue has an incomplete auto code gen for the CEF C++ headers where it doesn't correctly grab the base class. For example it only supports auto gen for classes implementing `CefBaseRefCounted` and `CefBaseScoped` while the views feature may implement `CefView` which implements `CefBaseRefCounted`. The python script doesn't know how to handle generating this correctly.

If you don't want to go down this route the only other option to explore would be figuring out how views work and seeing what can be done with the code we have. The main thing we're looking for is that the parent (our base window hosting the browser process) keeps all window events to itself and passes them down as necessary to the browser window.

**Comment by mattkol** [comment]
@NovusTheory..oh no.I did not really mean that I don't want to cross the boundary. As long as we have the knowledge to do it and it is easily maintainable, I am all for it. I am only being precautional so that we do not get into endless loop.

If we have the knowledge and we can continue to support it. We should be fine. 

Thanks.

**Comment by NovusTheory** [comment]
So @mattkol I may have come up with a good solution that won't cause problems and overall prevent lot of hacks. Firstly, I did try to get views implemented in CefGlue but the interop generator it has is a mess and I didn't want to spend the time figuring out CefGlue and and the interop generator.

My solution was to set the CEF browser as windowless. Instead of CEF creating a window and Chromely creating its own window to then have the browser as a child of it. I figured I'd strip CEF's window out and we just use our window.

How does this all work though? I use CefRenderHandler which exposes the renderer for CEF when using windowless mode. The main thing we want is `OnPaint` which provides a buffer of the pixels (BGRA array) that CEF would have otherwise natively rendered onto its own browser window by itself. What my implementation does, which currently only applies to CefGlue.WinApi, is I use GDI to create a bitmap from that pixel array, provide it to the NativeWindow, and call a window redraw which sends a WM_PAINT event to our Chromely window. In that WM_PAINT event I take the bitmap and do some conversion and interaction with HDC to draw it on the window client area.

As for interaction with that browser I use more WndProc events to pick up focus events, mouse scrolls, and key presses and send that off to the CefBrowserHost.

This implementation is most certainly not perfect right now but I want to get some feedback if this is a good direction to go into. The main reason behind doing this is that we don't have to do trickery with the CEF browser window and we instead now keep all WndProc events for ourselves and this makes implementing frameless and handling input much easier whereas before I had to override CEF's browser windows WndProcs to get what was needed to do this and input seemed to have been buggy.

The branch with this implementation is located here https://github.com/NovusTheory/Chromely/tree/impl/os.

**Comment by amaitland** [comment]
For what it's worth,  and obviously you can completely disregard this if you so choose.  I would advise against implementing OffScreen rendering,  you will gain some benefit,  the downsides outweigh those benefits. 

- Performance is worse
- There are constantly rendering issues which change from version to version 
- You have to implement your own DPI scaling, tooltips/popups, accessibility handling,  IME support,  the list goes on.
- From memory there is still an issue with adding custom context menu items

With more time I'm sure I could list more.

**Comment by mattkol** [comment]
@amaitland thanks for your input on this and support for Chromely.

@NovusTheory one major reason for Chromely from day one was to avoid Offscreen rendering. This I stated in an  earlier article on CodeProject:

https://www.codeproject.com/Articles/1243956/Chromely-Lightweight-Alternative-to-Electron

_"While working on a WPF project, SugarDesk, I was looking for ways to include SugarCrm Reports in my project and the most appealing option was to use Electron. Then, I discovered CefSharp but was not happy with the limitation - you can only use bitmap rendering instead of pure HTML rendering. In short, I ended up creating one without WPF and decided to share it with the community."_

From my research so far I think our best bet is still the Winapi Subclassing we started with. I have a skeletal implementation that we can fine tune (for Windows), but just not having the bandwidth yet to zoom in on it. Hopefully next couple of weekends.

Great efforts on this!

**Comment by mattkol** [comment]
@NovusTheory @FrankPfattheicher  using Subclassing, this is now implemented for Windows. The implementation is based on CefSharp [ChromeWidgetMessageInterceptor](https://github.com/cefsharp/CefSharp/blob/master/CefSharp.WinForms.Example/ChromeWidgetMessageInterceptor.cs).

Main implementation features are:
1. [Chromely implementation of ChromeWidgetMessageInterceptor](https://github.com/chromelyapps/Chromely/blob/master/src/Common/ChromeWidgetMessageInterceptor.cs).
2. [Registration of the interceptor](https://github.com/chromelyapps/Chromely/blob/dec7dd5a1a5bf014ea34a173daa97ad5bb898054/src/Chromely.CefSharp.Winapi/BrowserWindow/Window.cs#L153)
3. Handling of drag messages .. [Dragging](https://github.com/chromelyapps/Chromely/blob/dec7dd5a1a5bf014ea34a173daa97ad5bb898054/src/Chromely.CefSharp.Winapi/BrowserWindow/Window.cs#L160)
4. Handling related NC related windows messages - [NC Messages](https://github.com/chromelyapps/Chromely/blob/dec7dd5a1a5bf014ea34a173daa97ad5bb898054/src/Common/NativeWindowBase.cs#L417)

I checked this into the master so that it is easier to test and discuss. And also easy to revert if need be.

There is a wiki page explaining what is required to for usage - 
https://github.com/chromelyapps/Chromely/wiki/Frameless-Window

**Comment by NovusTheory** [comment]
Maybe the communication was just getting mixed up in my head. I properly tested what was necessary for checking the subclassing and I noticed that every other LBUTTONDOWN does not get captured. It instead seems to be captured by Chromium as I noticed every other LBUTTONDOWN and dragging my cursor would highlight text in the window. Other than that it seems to work as it should and the code looks clean.

I also noticed you're handling NCCALCSIZE as defined. here https://github.com/chromelyapps/Chromely/blob/dec7dd5a1a5bf014ea34a173daa97ad5bb898054/src/Common/NativeWindowBase.cs#L440-L447 Is this intended behavior? Usually frameless windows have no non client area and just `return IntPtr.Zero`.

**Comment by mattkol** [comment]
@NovusTheory 

While I was implementing this, Subclassing with dragging seems to work great. All the time ..

The resizing was the annoying part (or I lost patience at the end, or just dumb!).

So I think this is where your expertise will help - using your code on resizing. 

The problem is when I tried the usual process of NC* hit tests and returning appropriate flags - HTTOPLEFT, HTBOTTOMRIGHT, etc .. I lost the subclassing working appropriately.

So I tried using - WS_SIZEBOX -
https://github.com/chromelyapps/Chromely/blob/b424c8a370542e5e508d13a3d2db2245c44a8095/src/Common/NativeWindowBase.cs#L514
 
This got resizing to work well. 

But also got the [WhiteStrip](https://stackoverflow.com/questions/39731497/create-window-without-titlebar-with-resizable-border-and-without-bogus-6px-whit)

So what you see is trying to get around the whitestrip. Ugly yes, but I still rather have the subclassing and dragging working well.

Please throw that out, if you find alternatives.

**Comment by NovusTheory** [comment]
@mattkol After some testing I figure out why LBUTTONDOWN was not consistently working. When you define the WndProc here https://github.com/chromelyapps/Chromely/blob/b424c8a370542e5e508d13a3d2db2245c44a8095/src/Common/ChromeWidgetMessageInterceptor.cs#L99-L104 you are not returning `IntPtr.Zero` (what windows requires when an application handles the message) and so it falls out of the switch case and sends back the result of this https://github.com/chromelyapps/Chromely/blob/b424c8a370542e5e508d13a3d2db2245c44a8095/src/Common/ChromeWidgetMessageInterceptor.cs#L110 Simply placing a `return IntPtr.Zero` after `_forwardAction(message);` resulted in it working consistently every time. 

Was there a chance that when you tested it you were clicking the non client area as per your NCCALCSIZE you made. I removed the non client area when I was testing this.

If this seems to work consistently and doesn't cause problems with the subclassing then we could continue to the rest of the implementation.

**Comment by mattkol** [comment]
@NovusTheory makes sense on return IntPtr.Zero. I was returning the default handling result which was ruining everything. Returning IntPtr.Zero did not cross my mind.

I made the [change](https://github.com/chromelyapps/Chromely/commit/13a625ac4a24f6438caf25e4e47065c304744e80).

On NCCALCSIZE, it not that obvious if you use the usual white background. You have to use a different color to see it.
![image](https://user-images.githubusercontent.com/18384207/64957097-bb6dda00-d851-11e9-8087-afbf38adedb3.png)

Great progress!!!

**Comment by NovusTheory** [comment]
@mattkol You mentioned earlier that you had issues with NCHITTEST? I was doing some testing with it because I can't see resize borders being implemented correctly without it and I'd like you to try something out.
```csharp
case WM.NCHITTEST:
    var hitTest = GetCursorHitTest(hWnd, wParam, lParam);
    if (hitTest != HitTestResult.HTCLIENT)
    {
        return (IntPtr)HitTestResult.HTTRANSPARENT;
    }
    break;
```
Using some old code in the commits I did for hit test checking I'd like you to try to place the above code above or around this line inside the switch case. `GetCursorHitTest` is the old hit test code.
https://github.com/chromelyapps/Chromely/blob/b424c8a370542e5e508d13a3d2db2245c44a8095/src/Common/ChromeWidgetMessageInterceptor.cs#L91-L93

I set it up to replace the int[,] to HitTestResult[,] and replaced each one accordingly. The middle (a 0) being HTCLIENT. I set it up so that anything which *isn't* the client will be transparent and sent to the NativeWindowBase class while anything that is HTCLIENT will pass-through to chromium's wndproc. I setup NativeWindowBase NCHITTEST real quick to just `HTCAPTION` and it seems to work fine. I didn't notice any subclassing being broken and it shouldn't be if it's in the client area. If all goes well if you try this out it may be an option to continue with this and figure out the best code implementation.
